target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,461 @@
+{
+	return 1;
+}
+
+/*
+ * Initialization functions to register the clocks.
+ */
+
+static const struct clk_ops rtl838x_pll_ops = {
+	.round_rate = rtl838x_pll_round_rate,
+	.recalc_rate = rtl83xx_pll_recalc_rate,
+	.set_rate = rtl838x_pll_set_rate,
+	.enable = rtl838x_pll_enable,
+	.disable = rtl838x_pll_disable,
+	.is_enabled = rtl838x_pll_is_enabled
+};
+
+static const struct clk_ops rtl839x_pll_ops = {
+	.round_rate = rtl839x_pll_round_rate,
+	.recalc_rate = rtl83xx_pll_recalc_rate,
+	.set_rate = rtl839x_pll_set_rate,
+	.enable = rtl839x_pll_enable,
+	.disable = rtl839x_pll_disable,
+	.is_enabled = rtl839x_pll_is_enabled
+};
+
+static struct rtl_pll_data *rtl_pll_create_data(struct device_node *np)
+{
+	struct rtl_pll_data *data;
+
+	data = kzalloc(sizeof(*data), GFP_KERNEL);
+	if (!data)
+		return ERR_PTR(-ENOMEM);
+
+	data->np = np;
+
+	return data;
+}
+
+struct rtl_pll *rtl_pll_hw_register(const struct rtl_pll_init_data *pll_init)
+{
+	struct clk_parent_data parent_data = { };
+	struct clk_init_data hw_init = { };
+	struct rtl_pll *pll;
+	int ret;
+
+	if (!pll_init)
+		return ERR_PTR(-EINVAL);
+
+	pll = kzalloc(sizeof(*pll), GFP_KERNEL);
+	if (!pll)
+		return ERR_PTR(-ENOMEM);
+
+	pll->hw.init = &hw_init;
+	pll->id = pll_init->id;
+	spin_lock_init(&pll->lock);
+
+	hw_init.name = pll_init->name;
+	hw_init.ops = pll_init->ops;
+
+	if (!pll_init->parent_name) {
+		ret = -EINVAL;
+		goto err_free_pll;
+	}
+	parent_data.fw_name = pll_init->parent_name;
+	hw_init.parent_data = &parent_data;
+	hw_init.num_parents = 1;
+
+	ret = of_clk_hw_register(pll_init->np, &pll->hw);
+	if (ret)
+		goto err_free_pll;
+
+	clk_hw_register_clkdev(&pll->hw, pll_init->name, NULL);
+	return pll;
+
+err_free_pll:
+	kfree(pll);
+
+	return ERR_PTR(ret);
+}
+
+static struct rtl_pll *rtl_pll_find_desc(struct rtl_pll_data *data, unsigned int clk_id)
+{
+	struct rtl_pll *pll;
+	int idx;
+
+	for (idx = 0; idx < PLL_COUNT; ++idx) {
+		pll = data->plls[idx];
+		if (pll && pll->id == clk_id)
+			return pll;
+	}
+
+	return ERR_PTR(-EINVAL);
+}
+
+static struct clk_hw *rtl_pll_of_clk_hw_get(struct of_phandle_args *clkspec, void *priv)
+{
+	struct rtl_pll_data *data = priv;
+	struct rtl_pll *pll;
+	unsigned int clk_id;
+
+	clk_id = clkspec->args[0];
+	pll = rtl_pll_find_desc(data, clk_id);
+	if (IS_ERR(pll)) {
+		pr_info("Invalid PLL clock ID %d specified\n", clk_id);
+		return ERR_CAST(pll);
+	}
+
+	return pll ? &pll->hw : NULL;
+}
+
+void rtl_pll_hw_unregister(struct rtl_pll *pll)
+{
+	clk_hw_unregister(&pll->hw);
+
+	kfree(pll);
+}
+
+static int rtl_pll_clk_register(struct rtl_pll_data *data)
+{
+	int idx, ret;
+
+	for (idx = 0; idx < PLL_COUNT; ++idx) {
+		const struct rtl_pll_info *info = &pll_info[idx];
+		struct rtl_pll_init_data init = {0};
+
+		init.id = info->id;
+		init.name = info->name;
+		init.parent_name = info->parent_name;
+		init.np = data->np;
+		if (rtl_soc_family == RTL8390_FAMILY_ID)
+			init.ops = &rtl839x_pll_ops;
+		else
+			init.ops = &rtl838x_pll_ops;
```

Comment:
as before, make it explicit and error out in the other case.

I wonder if we can combine the RTL930x in this driver as well; shouldn't be that unreasonable. I think the RTL931x might be tricky, as it's a different mips core; but even that might feature many similarities ...

Reply:
rewritten